File: src/scripts/protocols/lib/data-writer.ts
export class DataWriter {
  private _buf: Buffer
  public pos = 0

  constructor() {
    this._buf = new Buffer(256)
  }

  get buf() {
    return this._buf.slice(0, this.pos)
  }

  u8(val: number) {
    this._buf.writeUInt8(val, this.pos)
    this.pos += 1
    return this
  }

  u16le(val: number) {
    this._buf.writeUInt16LE(val, this.pos)
    this.pos += 2
    return this
  }

  u32le(val: number) {
    this._buf.writeUInt32LE(val, this.pos)
    this.pos += 4
    return this
  }

  i8(val: number) {
    this._buf.writeInt8(val, this.pos)
    this.pos += 1
    return this
  }

  i16le(val: number) {
    this._buf.writeInt16LE(val, this.pos)
    this.pos += 2
    return this
  }

  i32le(val: number) {
    this._buf.writeInt32LE(val, this.pos)
    this.pos += 4
    return this
  }

  f32le(val: number) {
    this._buf.writeFloatLE(val, this.pos)
    this.pos += 4
    return this
  }

  f64le(val: number) {
    this._buf.writeDoubleLE(val, this.pos)
    this.pos += 8
    return this
  }

  zstring(val: string) {
    this.string(val, val.length)
    this.u8(0)
    return this
  }

  string(val: string, length: number = val.length) {
    this._buf.write(val, this.pos, length)
    this.pos += length
    return this
  }

  data(data: Buffer) {
    this._buf = Buffer.concat([this.buf, data])
    this.pos = this._buf.length
    return this
  }
}

File: src/protocols/lib/DataWriter.lua
local Object = require("lib/classic")

local DataWriter = Object:extend()

function DataWriter:new()
  self.buf = ""
end

function DataWriter:u8(val)
  self.buf = self.buf .. string.pack("I1", val)
  return self
end

function DataWriter:u16le(val)
  self.buf = self.buf .. string.pack("<I2", val)
  return self
end

function DataWriter:u32le(val)
  self.buf = self.buf .. string.pack("<I4", val)
  return self
end

function DataWriter:u64le(val)
  self.buf = self.buf .. string.pack("<I8", val)
  return self
end

function DataWriter:i8(val)
  self.buf = self.buf .. string.pack("i1", val)
  return self
end

function DataWriter:i16le(val)
  self.buf = self.buf .. string.pack("<i2", val)
  return self
end

function DataWriter:i32le(val)
  self.buf = self.buf .. string.pack("<i4", val)
  return self
end

function DataWriter:i64le(val)
  self.buf = self.buf .. string.pack("<i8", val)
  return self
end

function DataWriter:f32le(val)
  self.buf = self.buf .. string.pack("<f", val)
  return self
end

function DataWriter:f64le(val)
  self.buf = self.buf .. string.pack("<d", val)
  return self
end

function DataWriter:zstring(val)
  self.buf = self.buf .. string.pack("z", val)
  return self
end

function DataWriter:string(val, len)
  self.buf = self.buf .. val:sub(1, len or #val)
  return self
end

function DataWriter:data(val, len)
  self.buf = self.buf .. val:sub(1, len or #val)
  return self
end

return DataWriter
